File: python_solutions/chapter_05_bit_manipulation/problem_05_02_binary_to_string.py
def binary_to_string(fraction):
    if fraction >= 1 or fraction <= 0:
        return None
    binary_string = '0.'
    i = 0
    while i < 32 and fraction > 0:  # assume our algorithm can only compute the first 32 bits of the fractional value
        if fraction * 2 >= 1:
            binary_string += '1'
            fraction = fraction * 2 - 1
        else:
            binary_string += '0'
            fraction *= 2
        i += 1
    return binary_string

File: cpp_solutions/chapter_05_bit_manipulation/problem_05_02_binaryToString.cpp
#include "problem_05_02_binaryToString.h"

namespace chapter_05{
    int binaryToString(double fraction, std::string& binaryString){
        if (fraction >= 1 || fraction <= 0){
            return -1;
        }
        binaryString = "0.";
        int i = 0;
        while (fraction > 0.0 && i < 32){
            if (fraction * 2.0 >= 1.0){
                binaryString.append("1");
                fraction = fraction * 2.0 - 1.0;
            } else {
                binaryString.append("0");
                fraction *= 2.0;
            }
            i ++;
        }
        return 0;
    }
}

